service/matching/tasklist/adaptive_scaler.go

```diff
@@ -177,91 +192,149 @@ func (a *adaptiveScalerImpl) getPartitionConfig() *types.TaskListPartitionConfig
 	partitionConfig := a.tlMgr.TaskListPartitionConfig()
 	if partitionConfig == nil {
 		partitionConfig = &types.TaskListPartitionConfig{
-			NumReadPartitions:  1,
-			NumWritePartitions: 1,
+			ReadPartitions: map[int]*types.TaskListPartition{
+				0: {},
+			},
+			WritePartitions: map[int]*types.TaskListPartition{
+				0: {},
+			},
 		}
 	}
 	return partitionConfig
 }
 
-func (a *adaptiveScalerImpl) adjustWritePartitions(qps float64, numWritePartitions int32) int32 {
-	upscaleThreshold := float64(a.config.PartitionUpscaleRPS())
+func (a *adaptiveScalerImpl) calculateWritePartitionCount(qps float64, numWritePartitions int) int {
+	upscaleRps := float64(a.config.PartitionUpscaleRPS())
+	partitions := float64(numWritePartitions)
 	downscaleFactor := a.config.PartitionDownscaleFactor()
-	downscaleThreshold := float64(numWritePartitions-1) * upscaleThreshold * downscaleFactor / float64(numWritePartitions)
+	upscaleThreshold := partitions * upscaleRps
+	downscaleThreshold := (partitions - 1) * upscaleRps * downscaleFactor
 	a.scope.UpdateGauge(metrics.EstimatedAddTaskQPSGauge, qps)
 	a.scope.UpdateGauge(metrics.TaskListPartitionUpscaleThresholdGauge, upscaleThreshold)
 	a.scope.UpdateGauge(metrics.TaskListPartitionDownscaleThresholdGauge, downscaleThreshold)
 
 	result := numWritePartitions
-	if qps > upscaleThreshold {
-		if !a.overLoad {
-			a.overLoad = true
-			a.overLoadStartTime = a.timeSource.Now()
-		} else if a.timeSource.Now().Sub(a.overLoadStartTime) > a.config.PartitionUpscaleSustainedDuration() {
-			result = getNumberOfPartitions(numWritePartitions, qps, upscaleThreshold)
-			a.logger.Info("adjust write partitions", tag.CurrentQPS(qps), tag.PartitionUpscaleThreshold(upscaleThreshold), tag.PartitionDownscaleThreshold(downscaleThreshold), tag.PartitionDownscaleFactor(downscaleFactor), tag.CurrentNumWritePartitions(numWritePartitions), tag.NumWritePartitions(result))
-			a.overLoad = false
-		}
-	} else {
-		a.overLoad = false
+	if a.overLoad.Check(qps > upscaleThreshold) {
```

Comment:
this doesn't seem correct to me. Originally, we're comparing the load of a single partition with the threshold. Now, we're comparing the total load of all partitions with the threashold.